x-pack/plugins/siem/server/lib/detection_engine/signals/filter_events_with_list.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { get } from 'lodash/fp';
+import { Logger } from 'src/core/server';
+
+import { ListItemArraySchema } from '../../../../../lists/common/schemas/response';
+import { Type as ListValueType } from '../../../../../lists/common/schemas/common';
+import { ListClient } from '../../../../../lists/server';
+import { SignalSearchResponse } from './types';
+import { RuleAlertParams } from '../types';
+
+interface FilterEventsAgainstList {
+  listClient: ListClient;
+  exceptionsList: RuleAlertParams['exceptions_list'];
+  logger: Logger;
+  eventSearchResult: SignalSearchResponse;
+}
+
+export const filterEventsAgainstList = async ({
+  listClient,
+  exceptionsList,
+  logger,
+  eventSearchResult,
+}: FilterEventsAgainstList): Promise<SignalSearchResponse> => {
+  try {
+    if (exceptionsList == null) {
+      return eventSearchResult;
+    }
+
+    // grab the signals with values found in the given exception lists.
+    const filteredHitsPromises = exceptionsList.map(async exceptionItem => {
+      // acquire the list values we are checking for.
+      const valuesOfGivenType = eventSearchResult.hits.hits
+        .map(searchResultItem => get(exceptionItem.field, searchResultItem._source))
+        .filter((searchResultItem: string) => searchResultItem != null);
+      const valueSet = new Set<string>(valuesOfGivenType); // make them small
+      if (exceptionItem.values == null || exceptionItem.values.length === 0) {
+        throw new Error('Malformed exception list provided');
+      }
+      const listSignals: ListItemArraySchema = await listClient.getListItemByValues({
+        listId: exceptionItem.values[0].id!,
+        type: exceptionItem.values[0].name as ListValueType, // bad, I know, I'll write a typeguard later.
+        value: [...valueSet],
+      });
+      // create a set of list values that were a hit
+      const badSet = new Set<string>(listSignals.map(item => item.value));
+      // console.log({ badSet });
+      // filter out the search results that match with the values found in the list.
+      const filteredEvents = eventSearchResult.hits.hits.filter(item =>
+        get(exceptionItem.field, item._source)
+          ? !badSet.has(get(exceptionItem.field, item._source))
+          : true
+      );
```

Comment:
I would opt for an `as cast` like so:

```ts
      const filteredEvents = eventSearchResult.hits.hits.filter(item =>
        item._source[exceptionItem.field] != null
          ? !badSet.has(item._source[exceptionItem.field] as string)
          : true
      );
```

Notice that I am checking against `!= null` though. With the way it is written you are going to match as false would be things like 0 and empty string:

```ts
0 ? true : false
false
1 ? true : false
true
```

With type narrowing and unknown you can do something like this:

```ts
      const filteredEvents = eventSearchResult.hits.hits.filter(item => {
        const field = item._source[exceptionItem.field];
        if (typeof field === 'string') {
          return !badSet.has(field);
        } else {
          return true;
        }
      });
```
which should handle all your use cases.

Reply:
I added this function

```node.js
// narrow unioned type to be single
const isStringableType = (val: SearchTypes) =>
  ['string', 'number', 'boolean'].includes(typeof val);
```

which should do the narrowing, and then I just call `.toString()` once I know that the value is not null.

Reply:
Fixed [here](6ebc0cdd8397706dd57c45570661610de85d7de1)